File: algorithms/java/src/test/java/leetcode/LC107BinaryTreeLevelOrderTraversalII.java
package leetcode.lc107_binary_tree_level_order_traversal_i_i;

import org.junit.jupiter.api.Test;


import java.util.ArrayList;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;

/**
 * https://leetcode.com/problems/binary-tree-level-order-traversal-ii/
 * #medium
 */
public final class LC107BinaryTreeLevelOrderTraversalII {
    public final class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(final int x) {
            val = x;
        }
    }

    public final class Solution {
        public List<List<Integer>> levelOrderBottom(final TreeNode root) {
            List<List<Integer>> result = new ArrayList<>();
            List<TreeNode> current = new ArrayList<>();
            if (root != null) {
                current.add(root);
            }
            while (current.size() > 0) {
                List<Integer> level = new ArrayList<>();
                result.add(level);
                List<TreeNode> future = new ArrayList<>();
                for (TreeNode node : current) {
                    level.add(node.val);
                    if (node.left != null) {
                        future.add(node.left);
                    }
                    if (node.right != null) {
                        future.add(node.right);
                    }
                }
                current = future;
            }
            int length = result.size();
            for (int i = 0; i < length / 2; i++) {
                List<Integer> exchange = result.get(i);
                result.set(i, result.get(length - 1 - i));
                result.set(length - 1 - i, exchange);
            }
            return result;
        }
    }

    @Test
    public void test_empty() {
        assertEquals(0, new Solution().levelOrderBottom(null).size());
    }

    @Test
    public void test_example() {
        TreeNode n3 = new TreeNode(3);
        TreeNode n9 = new TreeNode(9);
        TreeNode n20 = new TreeNode(20);
        TreeNode n15 = new TreeNode(15);
        TreeNode n7 = new TreeNode(7);
        n3.left = n9;
        n3.right = n20;
        n20.left = n15;
        n20.right = n7;
        int[][] expected = {{15, 7}, {9, 20}, {3}};
        List<List<Integer>> result = new Solution().levelOrderBottom(n3);
        assertEquals(expected.length, result.size());
        for (int i = 0; i < expected.length; i++) {
            List<Integer> level = result.get(i);
            assertEquals(expected[i].length, level.size());
            for (int j = 0; j < expected[i].length; j++) {
                assertEquals(expected[i][j], level.get(j).intValue());
            }
        }
    }
}

File: algorithms/code/leetcode/lc107_binary_tree_level_order_traversal_ii/lc107_binary_tree_level_order_traversal_ii.rs
// https://leetcode.com/problems/sum-root-to-leaf-numbers/
// #easy

use std::cell::RefCell;
use std::rc::Rc;

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    pub fn new(
        val: i32,
        left: Option<Rc<RefCell<TreeNode>>>,
        right: Option<Rc<RefCell<TreeNode>>>,
    ) -> Option<Rc<RefCell<Self>>> {
        Some(Rc::new(RefCell::new(TreeNode { val, left, right })))
    }
}

pub struct Solution;

impl Solution {
    pub fn level_order_bottom(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<Vec<i32>> {
        let mut result: Vec<Vec<i32>> = Vec::new();
        let mut current: Vec<Rc<RefCell<TreeNode>>> = Vec::new();
        if let Some(root) = root {
            current.push(root);
        }
        while !current.is_empty() {
            let mut level: Vec<i32> = Vec::new();
            let mut future: Vec<Rc<RefCell<TreeNode>>> = Vec::new();
            for node in &current {
                let node = node.borrow();
                level.push(node.val);
                if let Some(left) = &node.left {
                    future.push(left.clone());
                }
                if let Some(right) = &node.right {
                    future.push(right.clone());
                }
            }
            result.push(level);
            current = future;
        }
        result.reverse();
        result
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_empty() {
        assert!(Solution::level_order_bottom(None).is_empty());
    }

    #[test]
    fn test_example() {
        let n15 = TreeNode::new(15, None, None);
        let n7 = TreeNode::new(7, None, None);
        let n9 = TreeNode::new(9, None, None);
        let n20 = TreeNode::new(20, n15, n7);
        let n3 = TreeNode::new(3, n9, n20);
        let result = Solution::level_order_bottom(n3);
        let expected = vec![vec![15, 7], vec![9, 20], vec![3]];
        assert_eq!(expected, result);
    }
}
